File: totoggle_go/internal/toggle/application_test.go
package toggle

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func mustPath(t *testing.T, raw string) Path {
	t.Helper()
	p, err := NewPath(raw)
	require.NoError(t, err)
	return p
}

// The user's own t1.t2.t3 hierarchy example: querying the leaf must surface every ancestor on the
// way down from the root, root-first — this is the exact cascade Application.AncestorsOf owns so
// the client's hierarchical evaluation isn't reimplementing path-walking itself.
func TestApplication_AncestorsOf_ThreeLevelHierarchy(t *testing.T) {
	t1 := Toggle{ID: "1", Path: mustPath(t, "t1"), Enabled: true}
	t1t2 := Toggle{ID: "2", Path: mustPath(t, "t1.t2"), Enabled: true}
	t1t2t3 := Toggle{ID: "3", Path: mustPath(t, "t1.t2.t3"), Enabled: true}

	app := NewApplication([]Toggle{t1, t1t2, t1t2t3})

	ancestors := app.AncestorsOf(mustPath(t, "t1.t2.t3"))
	require.Len(t, ancestors, 2)
	assert.Equal(t, t1, ancestors[0])
	assert.Equal(t, t1t2, ancestors[1])
}

func TestApplication_AncestorsOf_RootHasNoAncestors(t *testing.T) {
	t1 := Toggle{ID: "1", Path: mustPath(t, "t1"), Enabled: true}
	app := NewApplication([]Toggle{t1})

	assert.Empty(t, app.AncestorsOf(mustPath(t, "t1")))
}

// A segment missing from the fetched set (never configured, or filtered out server-side) never
// affects the result — it's simply absent from the cascade rather than an error.
func TestApplication_AncestorsOf_SkipsMissingAncestors(t *testing.T) {
	t1 := Toggle{ID: "1", Path: mustPath(t, "t1"), Enabled: true}
	t1t2t3 := Toggle{ID: "3", Path: mustPath(t, "t1.t2.t3"), Enabled: true}
	app := NewApplication([]Toggle{t1, t1t2t3})

	ancestors := app.AncestorsOf(mustPath(t, "t1.t2.t3"))
	require.Len(t, ancestors, 1)
	assert.Equal(t, t1, ancestors[0])
}

func TestApplication_ByPath_Found(t *testing.T) {
	t1t2 := Toggle{ID: "2", Path: mustPath(t, "t1.t2"), Enabled: true}
	app := NewApplication([]Toggle{t1t2})

	found, ok := app.ByPath(mustPath(t, "t1.t2"))
	require.True(t, ok)
	assert.Equal(t, t1t2, found)
}

func TestApplication_ByPath_NotFound(t *testing.T) {
	app := NewApplication(nil)

	_, ok := app.ByPath(mustPath(t, "missing"))
	assert.False(t, ok)
}

File: totoggle_node/src/internal/toggle/application.test.ts
import { describe, expect, it } from "vitest";
import { Path } from "./path.js";
import type { Toggle } from "./toggle.js";
import { Application } from "./application.js";

function toggle(path: string, overrides: Partial<Toggle> = {}): Toggle {
  return {
    id: path,
    path: Path.parse(path),
    value: path,
    enabled: true,
    level: 0,
    parentId: null,
    appId: "app-1",
    hasActivationRule: false,
    activationRule: null,
    ...overrides,
  };
}

describe("Application", () => {
  // The user's own t1.t2.t3 hierarchy example: querying the leaf must surface every ancestor
  // root-first, on the way down from the root.
  it("returns ancestors root-first for a three-level hierarchy", () => {
    const t1 = toggle("t1");
    const t1t2 = toggle("t1.t2");
    const t1t2t3 = toggle("t1.t2.t3");
    const app = new Application([t1, t1t2, t1t2t3]);

    expect(app.ancestorsOf(Path.parse("t1.t2.t3"))).toEqual([t1, t1t2]);
  });

  it("a root toggle has no ancestors", () => {
    const app = new Application([toggle("t1")]);
    expect(app.ancestorsOf(Path.parse("t1"))).toEqual([]);
  });

  // A segment missing from the fetched set never affects the result — it's simply absent from
  // the cascade rather than an error.
  it("skips ancestors that were never fetched", () => {
    const t1 = toggle("t1");
    const t1t2t3 = toggle("t1.t2.t3");
    const app = new Application([t1, t1t2t3]);

    expect(app.ancestorsOf(Path.parse("t1.t2.t3"))).toEqual([t1]);
  });

  it("looks up a toggle by exact path", () => {
    const t1t2 = toggle("t1.t2");
    const app = new Application([t1t2]);

    expect(app.byPath(Path.parse("t1.t2"))).toEqual(t1t2);
  });

  it("returns undefined for a path that was never fetched", () => {
    const app = new Application([]);
    expect(app.byPath(Path.parse("missing"))).toBeUndefined();
  });
});
